tests/test_state.py

```diff
@@ -0,0 +1,243 @@
+        patch(f"{patch_path}.set_virtual_node_map") as mock_set_virtual_node_map,\
+        patch(f"{patch_path}.set_parent_virtual_node_map") as mock_set_parent_virtual_node_map:
+
+        ujt.state.add_virtual_node(virtual_node_name, selected_node_data)
+
+        set_virtual_node_map_args, _ = mock_set_virtual_node_map.call_args
+        assert len(set_virtual_node_map_args) == 1  # only one non-kw arg
+
+        new_virtual_node_map = set_virtual_node_map_args[0]
+        assert len(new_virtual_node_map) == 1  # only one item in dict
+
+        virtual_node = new_virtual_node_map[virtual_node_name]
+        assert virtual_node.name == virtual_node_name
+        expected_virtual_node_child_names = [
+            service_relative_names[0],
+            f"{service_relative_names[0]}.{endpoint_relative_names[0]}",
+            f"{service_relative_names[0]}.{endpoint_relative_names[1]}",
+        ]
+        assert_same_elements(
+            virtual_node.child_names,
+            expected_virtual_node_child_names)
+        assert virtual_node.collapsed
+        assert virtual_node.node_type == NodeType.NODETYPE_VIRTUAL
+
+        # ---
+
+        set_parent_virtual_node_map_args, _ = mock_set_parent_virtual_node_map.call_args
+        assert len(set_parent_virtual_node_map_args) == 1
+
+        new_parent_virtual_node_map = set_parent_virtual_node_map_args[0]
+        assert len(new_parent_virtual_node_map) == 3
+        for name in expected_virtual_node_child_names:  # careful! virtual node itself isn't added to parent map
+            assert name in new_parent_virtual_node_map
+            assert new_parent_virtual_node_map[name] == virtual_node_name
+
+
+def test_delete_virtual_node(patch_path):
+    child_names = ["child0", "child1", "child2"]
+    virtual_node_name = "virtual_node"
+    virtual_node_map = {
+        virtual_node_name:
+            VirtualNode(
+                name=virtual_node_name,
+                child_names=child_names,
+            ),
+    }
+    parent_virtual_node_map = {
+        child_name: virtual_node_name
+        for child_name in child_names
+    }
+
+    with patch(f"{patch_path}.get_virtual_node_map", Mock(return_value=virtual_node_map)),\
+        patch(f"{patch_path}.get_parent_virtual_node_map", Mock(return_value=parent_virtual_node_map)),\
+        patch(f"{patch_path}.set_virtual_node_map") as mock_set_virtual_node_map,\
+        patch(f"{patch_path}.set_parent_virtual_node_map") as mock_set_parent_virtual_node_map:
+
+        ujt.state.delete_virtual_node(virtual_node_name)
+
+        set_virtual_node_map_args, _ = mock_set_virtual_node_map.call_args
+        assert len(set_virtual_node_map_args) == 1
+
+        new_virtual_node_map = set_virtual_node_map_args[0]
+        assert len(new_virtual_node_map) == 0
+
+        # ---
+
+        set_parent_virtual_node_map_args, _ = mock_set_parent_virtual_node_map.call_args
+        assert len(set_parent_virtual_node_map_args) == 1
```

Comment:
why do we need to use intermediate variables like this?

Reply:
If we have a function like this:
```
def f(a, b=2):
    return a + b
```
The caller provide arguments like this:
```
x = f(1, b=3)
```
In this case b is a keyword argument (kwarg).
The mock's call_arg's property is a 2-tuple that provides (args, kwargs), so this line ensures that the mock was only provided a single non-kwarg.